eth/chains/base.py

```diff
@@ -315,13 +315,31 @@ def validate_gaslimit(self, header: BlockHeader) -> None:
     def validate_uncles(self, block: BaseBlock) -> None:
         raise NotImplementedError("Chain classes must implement this method")
 
-    @abstractmethod
+    @classmethod
     def validate_chain(
-            self,
+            cls,
             parent: BlockHeader,
             chain: Tuple[BlockHeader, ...],
             seal_check_random_sample_rate: int = 1) -> None:
-        raise NotImplementedError("Chain classes must implement this method")
+
+        all_indices = list(range(len(chain)))
+        if seal_check_random_sample_rate == 1:
+            headers_to_check_seal = set(all_indices)
+        else:
+            sample_size = len(all_indices) // seal_check_random_sample_rate
+            headers_to_check_seal = set(random.sample(all_indices, sample_size))
+
+        for i, header in enumerate(chain):
```

Comment:
I think we can clean this loop up by doing:

```python
for i, (parent, header) in enumerate(sliding_window(2, concatv([parent], chain))):
    ...
```

This gives us pairs of `parent/child` allowing us to eliminate the `parent` tracking.